plotly.js always gets bundled

## Expected Behavior
Plotly.js does not get bundled into the production build if you are using either the `PlotlyViaCDNModule` or the `PlotlyViaWindowModule`.

## Actual Behavior
Plotly.js always gets bundled into the chunk the module is imported in (usually main).
This only happens if you install angular-plotly via npm, but not with the source project.

## Steps to Reproduce the Problem
  1. Make a new angular project with angular-plotly.js and plotly.js (like described in the README.md)
  1. Add either of the two modules to the app.module.ts (like described in the README.md)
  1. Create a production build with `ng build --prod`
You will notice that plotly.js will be bundled into the `main.<hash>.js` bundle (file size ~3MB)

## Specifications
  - Version: 1.1.0

## Other Notes
As mentioned above this only happens if you install angular-plotly.js via npm.
If you download the sources (`plotly-via-cdn.module.ts` and the `shared/` folder) and add them manually everything works as intended.
I don't know enough about npm-packages, but it seems that bundling makes it so all Modules are imported into the application as soon as you import one.

I uninstalled plotly.js to see where it is being called and ran ng build and the error is `ERROR in ./node_modules/angular-plotly.js/fesm5/angular-plotly.js.js
Module not found: Error: Can't resolve 'plotly.js/dist/plotly.js' in 'E:\Docker\ngxp\src\budgetRoad\node_modules\angular-plotly.js\fesm5'`

I'm assuming it should be pointed at the plotly.min.js. I'll look into it some more

--- a/README.md
+++ b/README.md
@@ -43,7 +43,10 @@ Add the `PlotlyModule` into the main app module of your project
 import { NgModule } from '@angular/core';
 import { CommonModule } from '@angular/common';
 
+import * as PlotlyJS from 'plotly.js/dist/plotly.js';
 import { PlotlyModule } from 'angular-plotly.js';
+
+PlotlyModule.plotlyjs = PlotlyJS;
 
 @NgModule({
     imports: [CommonModule, PlotlyModule],

--- a/src/app/demo/demo.module.ts
+++ b/src/app/demo/demo.module.ts
@@ -3,6 +3,8 @@ import { CommonModule } from '@angular/common';
 import { HttpClientModule } from '@angular/common/http';
 import { RouterModule, Routes } from '@angular/router';
 
+import * as PlotlyJS from 'plotly.js/dist/plotly.js';
+
 import { HomeComponent } from './home/home.component';
 import { DemoComponent } from './demo.component';
 
@@ -33,12 +35,16 @@ const demoRoutes: Routes = [
     { path: '', redirectTo: '/home', pathMatch: 'full' },
 ];
 
+
+PlotlyModule.plotlyjs = PlotlyJS;
+// PlotlyViaCDNModule.plotlyVersion = 'latest';
+
 @NgModule({
     imports: [
         CommonModule,
         HttpClientModule,
         PlotlyModule,
-        // PlotlyViaCDNModule.forRoot({version: 'latest'}),
+        // PlotlyViaCDNModule,
         // PlotlyViaWindowModule,
         RouterModule.forRoot(demoRoutes, { enableTracing: true }),
     ],

--- a/src/app/plotly-via-cdn/plotly-via-cdn.module.ts
+++ b/src/app/plotly-via-cdn/plotly-via-cdn.module.ts
@@ -51,7 +51,7 @@ export class PlotlyViaCDNModule {
         fn();
     }
 
-    static forRoot(config: Partial<{version: string}>): ModuleWithProviders {
+    static forRoot(config: Partial<{version: string}>): ModuleWithProviders<PlotlyViaCDNModule> {
         if (config.version === undefined) {
             console.warn(`It's strongly recommended that you set a plotly version when using via CDN.`);
             config.version = 'latest';

--- a/src/app/plotly/plotly.module.spec.ts
+++ b/src/app/plotly/plotly.module.spec.ts
@@ -1,0 +1,25 @@
+import { PlotlyModule } from "./plotly.module";
+
+
+
+describe('PlotlyModule', () => {
+
+    it('should create', () => {
+        const fn = () => {
+            const mod = new PlotlyModule();
+        };
+
+        const msg = "Invalid PlotlyJS object. Please check https://github.com/plotly/angular-plotly.js#quick-start"
+                  + " to see how to add PlotlyJS to your project.";
+        expect(fn).toThrowError(msg);
+
+
+        const fn2 = () => {
+            PlotlyModule.plotlyjs = {plot: function() { }};
+            const mod = new PlotlyModule();
+        };
+
+        expect(fn2).not.toThrowError(msg);
+    });
+
+});

--- a/src/app/plotly/plotly.module.ts
+++ b/src/app/plotly/plotly.module.ts
@@ -12,7 +12,7 @@ import { SharedModule } from '../shared/shared.module';
     exports: [PlotComponent]
 })
 export class PlotlyModule {
-    public static plotlyjs: any;
+    public static plotlyjs: any = {};
 
     constructor() {
         if (!this.isValid()) {

--- a/src/app/plotly/plotly.module.ts
+++ b/src/app/plotly/plotly.module.ts
@@ -1,7 +1,5 @@
-import { NgModule, ModuleWithProviders } from '@angular/core';
+import { NgModule } from '@angular/core';
 import { CommonModule } from '@angular/common';
-
-import * as PlotlyJS from 'plotly.js/dist/plotly.js';
 
 import { PlotComponent } from '../shared/plot/plot.component';
 import { PlotlyService } from '../shared/plotly.service';
@@ -14,14 +12,20 @@ import { SharedModule } from '../shared/shared.module';
     exports: [PlotComponent]
 })
 export class PlotlyModule {
+    public static plotlyjs: any;
+
     constructor() {
-        PlotlyService.setPlotly(PlotlyJS);
+        if (!this.isValid()) {
+            const msg = "Invalid PlotlyJS object. Please check https://github.com/plotly/angular-plotly.js#quick-start"
+                      + " to see how to add PlotlyJS to your project.";
+            throw new Error(msg);
+        }
+
+        PlotlyService.setPlotly(PlotlyModule.plotlyjs);
     }
 
-    static forRoot(): ModuleWithProviders {
-        return {
-            ngModule: PlotlyModule,
-            providers: [PlotlyService]
-        };
+    private isValid(): boolean {
+        return PlotlyModule.plotlyjs !== undefined
+            && typeof PlotlyModule.plotlyjs.plot === 'function';
     }
 }

--- a/src/app/shared/plot/plot.component.spec.ts
+++ b/src/app/shared/plot/plot.component.spec.ts
@@ -141,6 +141,17 @@ describe('PlotComponent', () => {
         }, 13);
     });
 
+    it('should fail when plotlyInstance is undefined', () => {
+        component.plotlyInstance = undefined;
+
+        const error = new Error(`Plotly component wasn't initialized`);
+        const fn = () => {
+            component.updatePlot();
+        };
+
+        expect(fn).toThrow(error);
+    });
+
     it('should add handler into window.resize when useResizeHandler=true', () => {
         spyOn(component, 'getWindow').and.callFake(() => windowSpy);
 

